changelog/2128.removal.rst

```diff
@@ -0,0 +1,7 @@
+Standardized :meth:`~urllib3.response.HTTPResponse.read` to respect the semantics of BufferedIOBase regardless of compression. Specifically, this method:
+
+ * only returns an empty bytes object to indicate EOF (that is, the response has been fully consumed),
+ * never returns more bytes than requested,
+ * can issue multiple system calls.
+
+If you want to directly read the bytes as sent over the wire without all that logic, use :meth:`~urllib3.response.HTTPResponse.raw_read` instead.
```

Comment:
What's the use-case for the `raw_read` function on `HTTPResponse`? My thinking was the previous behavior of returning data while not following `io.BufferedIOBase.read` semantics was a bug.

Thinking on extensibility to HTTP/2: there's a chance that data is available in a buffer somewhere w/o needing a system call due to stream multiplexing, our `read` (and `readX`) functions should be documented to use either 0-1 (for `read1`, etc) or 0-N system calls instead of "1 or more".

Reply:
Thanks. I've opened https://github.com/urllib3/urllib3/issues/2769 to discuss this separately from this pull request. `raw_read` is now private again: [`341dca9` (#2712)](https://github.com/urllib3/urllib3/pull/2712/commits/341dca97ea666fb035017307979d781f4ccdbd0a)